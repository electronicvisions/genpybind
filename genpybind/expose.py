from __future__ import unicode_literals

import textwrap


from . import utils
from .registry import Registry
from .utils import quote

if False:  # pylint: disable=using-constant-test
    from .decls.declarations import Declaration  # pylint: disable=unused-import
    from typing import (  # pylint: disable=unused-import
        List, Optional, Text, Tuple, Union)


def expose_as(
        toplevel_declarations,  # type: List[Declaration]
        output_files,  # type: List[Text]
        module,  # type: Text
        doc=None,  # type: Optional[Text]
        isystem=None,  # type: Optional[List[Text]]
        includes=None,  # type: Optional[List[Text]]
        tags=None,  # type: Optional[List[Text]]
):
    # type: (...) -> Text
    tpl = textwrap.dedent("""
    #ifndef __GENPYBIND_GENERATED__
    #define __GENPYBIND_GENERATED__
    #endif

    #include <sstream>
    #include <pybind11/pybind11.h>
    #include <pybind11/stl.h>
    {isystem}

    {includes}

    namespace py = pybind11;

    template <typename T>
    struct genpybind_stringstream_helper
    {{
        std::string operator()(T const& value) const {{
            std::stringstream tmp;
            tmp << value;
            return tmp.str();
        }}
    }};

    // If `T` is registered with pybind11 the corresponding Python type is returned.
    // Else a warning is emitted and None is returned.
    template <typename T>
    py::object genpybind_get_type_object() {{
        std::type_info const &tp = typeid(T);
        auto tinfo = py::detail::get_type_info(tp, /*throw_if_missing=*/false);
        if (!tinfo) {{
            std::string name = tp.name();
            py::detail::clean_type_id(name);
            PyErr_WarnFormat(PyExc_Warning, /*stack_level=*/7,
                             "Reference to unknown type '%s'", name.c_str());
            return pybind11::none();
        }}
        return py::reinterpret_borrow<py::object>((PyObject *)tinfo->type);
    }}

    {functions}
    {function_declarations}

    PYBIND11_MODULE({module}, {var}) {{
    {var}.doc() = {doc};
    {statements}
    }}
    """).strip()

    split_tpl = textwrap.dedent("""
    #ifndef __GENPYBIND_GENERATED__
    #define __GENPYBIND_GENERATED__
    #endif

    #include <sstream>
    #include <pybind11/pybind11.h>
    #include <pybind11/stl.h>
    {isystem}

    {includes}

    namespace py = pybind11;

    template <typename T>
    struct genpybind_stringstream_helper
    {{
        inline std::string operator()(T const& value) const {{
            std::stringstream tmp;
            tmp << value;
            return tmp.str();
        }}
    }};

    template <typename T>
    inline py::object genpybind_get_type_object()
    {{
        auto tinfo = py::detail::get_type_info(
            typeid(T), /*throw_if_missing=*/true);
        return py::reinterpret_borrow<py::object>((PyObject*)tinfo->type);
    }}

    {functions}
    """).strip()

    var = "m"
    registry = Registry(tags=tags)

    statements = []  # type: List[Text]
    pending_declarations = []  # type: List[Tuple[Declaration, Text]]
    postamble_declarations = []  # type: List[Tuple[Declaration, Text]]
    functions = []
    functions_prefix= []

    def handle_return(declaration, value, postamble_only=False):
        # type: (Declaration, Union[None, Text, Tuple[Declaration, Text]], bool) -> None
        """
        expose() may either yield a string containing a statement to be emitted or
        a pair of a declaration and the identifier of the parent scope to use.
        In the latter case it will be added to a list of pending declarations on which
        expose_later() is called after all other normal declarations have been processed.
        If in addition the returned declaration has a truthy "postamble" attribute, it
        will be processed even later.
        """
        if value is None:
            return
        if not isinstance(value, tuple):
            assert utils.is_string(value)
            statements.append(value)
            return
        declaration, _ = value
        is_postamble = getattr(declaration, "postamble", None)
        if postamble_only or is_postamble:
            postamble_declarations.append(value)
            return
        pending_declarations.append(value)

    for declaration in toplevel_declarations:
        for value in utils.flatten(declaration.expose(var, registry)):
            handle_return(declaration, value)
        function = "\n".join(statements)
        if function.strip():
            functions.append(function)
        statements = []

    while pending_declarations:
        declaration, parent = pending_declarations.pop(0)
        for value in utils.flatten(declaration.expose_later(var, parent, registry)):
            handle_return(declaration, value)

    while postamble_declarations:
        declaration, parent = postamble_declarations.pop(0)
        for value in utils.flatten(declaration.expose_later(var, parent, registry)):
            handle_return(declaration, value, postamble_only=True)

    def find_recursive_attribute(name, lines, parent_names = []):
        for line in lines:
            # find definition
            if line.startswith("auto {}".format(name)):
                args = line.split('=')[1].split('(')[1].split(')')[0].split(', ')
                if args[0] == var:
                    # found outmost parent
                    return [args[1]] + parent_names
                else:
                    # not yet found outmost parent
                    return find_recursive_attribute(args[0], lines, [args[1]] + parent_names)
        raise Exception("Did not find definition of {}".format(name))

    def is_attr(line):
      return line.split('.')[1].startswith("attr")

    remaining_statements = []
    functions_appends = ["" for i in range(len(functions))]
    for statement in statements:
        lines = statement.split('\n')
        remaining_lines = []
        for line in lines:
            name = line.split('.')[0]
            if line.startswith("genpybind"):
                if is_attr(line):
                    # attr can't be reordered, make it use the module to get the object
                    parent_names = find_recursive_attribute(name, ("\n".join(functions)).split("\n"))
                    split = line.split('.')
                    split[0] = var + "".join([".attr({})".format(attr) for attr in parent_names])
                    remaining_lines.append(".".join(split))
                else:
                    found = False
                    # non-attr can be reordered into function defining symbol, do it
                    for f, function in enumerate(functions):
                        if function.find(name) != -1:
                            functions_appends[f] += line + "\n"
                            found = True
                            break
                    if not found:
                        remaining_lines.append(line)
            else:
                remaining_lines.append(line)
        remaining_statements.append("\n".join(remaining_lines))

    functions = [functions[i] + "\n" + functions_appends[i] for i in range(len(functions))]

    # find undefined py::module uses
    # first find all module definitions, record and remove them
    modules = []
    module_names = []
    module_definitions = []
    for f in range(len(functions)):
        for line in functions[f].split('\n'):
            if line.startswith("py::module"):
                modules.append(line.split(' ')[1])
                module_names.append(line.split('"')[1])
                module_definitions.append(line)
                functions[f] = functions[f].replace(line, "")

    # prepend dynamic acquisition of used but not defined module
    for m, name in zip(modules, module_names):
        for f in range(len(functions)):
            if functions[f].find(m) != -1:
                functions[f] = "auto {} = static_cast<py::module>({}.attr(\"{}\"));\n".format(
                    m, var, name) + functions[f]


    function_def = "void function_{}(py::module& {})\n{{\n{}\n}}\n"
    function_decl = "void function_{}(py::module&);"
    function_call = "function_{}({});"

    functions = [function_def.format(i, var, functions[i]) for i in range(len(functions))]
    remaining_statements = [function_call.format(i, var) for i in range(len(functions))] + \
        remaining_statements
    function_declarations = [function_decl.format(i) for i in range(len(functions))]

    # split functions such that the linecount-difference is smaller than max(linecounts)
    functions_linecounts = [f.count('\n') for f in functions]
    if len(functions): # FIXME: needs redesign, as not pretty
        functions_linecounts, functions = map(list, zip(*sorted(zip(functions_linecounts, functions))))
    split_functions = [[] for i in range(len(output_files))]
    split_functions_linecounts = [0] * len(output_files)
    for function, function_linecount in zip(reversed(functions), reversed(functions_linecounts)):
        to_index = split_functions_linecounts.index(min(split_functions_linecounts))
        split_functions[to_index].append(function)
        split_functions_linecounts[to_index] += function_linecount
    # sort split_functions in order to have PYBIND11_MODULE in first file with minimal line count
    split_functions_linecounts, split_functions = map(list, zip(*sorted(zip(split_functions_linecounts, split_functions))))

    # prepend remaining statements with module definitions
    remaining_statements = module_definitions + remaining_statements

    print(tpl.format(
        module=module,
        name=quote(module),
        doc=quote(doc),
        isystem="\n".join('#include <{}>'.format(f) for f in isystem or []),
        includes="\n".join('#include {}'.format(quote(f)) for f in includes or []),
        var=var,
        functions="\n".join(split_functions[0]),
        function_declarations="\n".join(function_declarations),
        statements="\n".join(remaining_statements)
    ), file=open(output_files[0], 'w'))

    for i, output_file in enumerate(output_files[1:]):
        print(split_tpl.format(
            isystem="\n".join('#include <{}>'.format(f) for f in isystem or []),
            includes="\n".join('#include {}'.format(quote(f)) for f in includes or []),
            functions="\n".join(split_functions[i + 1])
        ), file=open(output_file, 'w'))
