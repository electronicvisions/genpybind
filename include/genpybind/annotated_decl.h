#pragma once

#include "genpybind/annotations/annotation.h"

#include <clang/AST/Decl.h>
#include <clang/AST/DeclCXX.h>
#include <llvm/ADT/DenseMap.h>
#include <llvm/ADT/Optional.h>
#include <llvm/ADT/SmallPtrSet.h>
#include <llvm/ADT/SmallSet.h>
#include <llvm/ADT/SmallVector.h>
#include <llvm/ADT/StringRef.h>
#include <llvm/Support/Casting.h>

#include <functional>
#include <memory>
#include <string>
#include <utility>
#include <vector>

namespace clang {
class CXXConstructorDecl;
class CXXMethodDecl;
class LambdaExpr;
} // namespace clang

namespace genpybind {

bool hasAnnotations(const clang::Decl *decl, bool allow_empty = true);
llvm::SmallVector<llvm::StringRef, 1>
getAnnotationStrings(const clang::Decl *decl);

/// Return spelling derived from the name of the identifier that names
/// the declaration.
std::string getSpelling(const clang::NamedDecl *decl);

class AnnotatedDecl {
public:
  enum class Kind {
    Named,

    Namespace,
    Enum,
    Class,
    TypeAlias,
    Variable,

    Function,
    ConversionFunction,
    Method,
    Operator,
    Constructor,
    lastFunction = Constructor,
    lastNamed = Constructor,
  };

private:
  Kind kind;

public:
  AnnotatedDecl(Kind kind) : kind(kind) {}
  virtual ~AnnotatedDecl() = default;

  static std::unique_ptr<AnnotatedDecl> create(const clang::NamedDecl *decl);

  /// Process the given annotation and emit potential errors
  /// via the declarations associated diagnostics engine.
  /// \return whether the annotation has been processed (even with errors)
  virtual bool processAnnotation(const clang::Decl *decl,
                                 const annotations::Annotation &annotation) = 0;

  Kind getKind() const { return kind; };

  /// Retrieve annotations from the annotation attributes of the declaration.
  /// Any errors (e.g. invalid annotations) are reported via the declaration's
  /// associated diagnostics engine.
  void processAnnotations(const clang::Decl *decl);
};

class AnnotatedNamedDecl : public AnnotatedDecl {
protected:
  AnnotatedNamedDecl(Kind kind) : AnnotatedDecl(kind) {}

public:
  /// Alternative identifier to use when exposing the declaration.
  std::string spelling;
  /// Explicitly specified visibility of the declaration.
  /// If this is `None` the effective visibility will be determined
  /// based on the visibility of the parent declaration.
  llvm::Optional<bool> visible;

  AnnotatedNamedDecl() : AnnotatedDecl(Kind::Named) {}

  bool processAnnotation(const clang::Decl *decl,
                         const annotations::Annotation &annotation) override;

  static bool classof(const AnnotatedDecl *decl) {
    return classofKind(decl->getKind());
  }
  static constexpr bool classofKind(Kind kind) {
    return kind >= Kind::Named && kind <= Kind::lastNamed;
  }
};

class AnnotatedNamespaceDecl : public AnnotatedNamedDecl {
public:
  bool module = false;
  std::vector<std::string> only_expose_in;

  AnnotatedNamespaceDecl() : AnnotatedNamedDecl(Kind::Namespace) {}

  bool processAnnotation(const clang::Decl *decl,
                         const annotations::Annotation &annotation) override;

  static bool classof(const AnnotatedDecl *decl) {
    return classofKind(decl->getKind());
  }
  static constexpr bool classofKind(Kind kind) {
    return kind == Kind::Namespace;
  }
};

class AnnotatedEnumDecl : public AnnotatedNamedDecl {
public:
  bool arithmetic = false;
  llvm::Optional<bool> export_values;

  AnnotatedEnumDecl() : AnnotatedNamedDecl(Kind::Enum) {}

  bool processAnnotation(const clang::Decl *decl,
                         const annotations::Annotation &annotation) override;

  static bool classof(const AnnotatedDecl *decl) {
    return classofKind(decl->getKind());
  }
  static constexpr bool classofKind(Kind kind) { return kind == Kind::Enum; }
};

class AnnotatedRecordDecl : public AnnotatedNamedDecl {
public:
  bool dynamic_attr = false;
  llvm::SmallPtrSet<const clang::TagDecl *, 1> hide_base;
  llvm::SmallPtrSet<const clang::TagDecl *, 1> inline_base;
  std::string holder_type;

  AnnotatedRecordDecl() : AnnotatedNamedDecl(Kind::Class) {}

  bool processAnnotation(const clang::Decl *decl,
                         const annotations::Annotation &annotation) override;

  static bool classof(const AnnotatedDecl *decl) {
    return classofKind(decl->getKind());
  }
  static constexpr bool classofKind(Kind kind) { return kind == Kind::Class; }
};

class AnnotatedTypedefNameDecl : public AnnotatedNamedDecl {
public:
  /// If the underlying type has default visibility, make it visible explicitly.
  bool encourage = false;
  /// Expose the underlying type at the location of the type alias instead.
  bool expose_here = false;
  std::vector<annotations::Annotation> annotations_to_propagate;

  AnnotatedTypedefNameDecl() : AnnotatedNamedDecl(Kind::TypeAlias) {}

  bool processAnnotation(const clang::Decl *decl,
                         const annotations::Annotation &annotation) override;

  /// Applies all annotations of this declaration to another declaration.
  /// This is used to propagate spelling and visibility in the case of
  /// "expose_here" type aliases.
  void propagateAnnotations(const clang::Decl *decl,
                            AnnotatedDecl &other) const;

  static bool classof(const AnnotatedDecl *decl) {
    return classofKind(decl->getKind());
  }
  static constexpr bool classofKind(Kind kind) {
    return kind == Kind::TypeAlias;
  }
};

class AnnotatedFunctionDecl : public AnnotatedNamedDecl {
protected:
  AnnotatedFunctionDecl(Kind kind) : AnnotatedNamedDecl(kind) {}

public:
  llvm::SmallVector<std::pair<unsigned, unsigned>, 1> keep_alive;
  llvm::SmallSet<unsigned, 1> noconvert;
  llvm::SmallSet<unsigned, 1> required;
  std::string return_value_policy;

  AnnotatedFunctionDecl() : AnnotatedNamedDecl(Kind::Function) {}

  bool processAnnotation(const clang::Decl *decl,
                         const annotations::Annotation &annotation) override;

  static bool classof(const AnnotatedDecl *decl) {
    return classofKind(decl->getKind());
  }
  static constexpr bool classofKind(Kind kind) {
    return kind >= Kind::Function && kind <= Kind::lastFunction;
  }
};

class AnnotatedConversionFunctionDecl final : public AnnotatedFunctionDecl {
public:
  AnnotatedConversionFunctionDecl()
      : AnnotatedFunctionDecl(Kind::ConversionFunction) {}

  static bool classof(const AnnotatedDecl *decl) {
    return classofKind(decl->getKind());
  }
  static constexpr bool classofKind(Kind kind) {
    return kind == Kind::ConversionFunction;
  }
};

class AnnotatedMethodDecl final : public AnnotatedFunctionDecl {
public:
  llvm::SmallSet<std::string, 1> getter_for;
  llvm::SmallSet<std::string, 1> setter_for;

  AnnotatedMethodDecl() : AnnotatedFunctionDecl(Kind::Method) {}

  bool processAnnotation(const clang::Decl *decl,
                         const annotations::Annotation &annotation) override;

  static bool classof(const AnnotatedDecl *decl) {
    return classofKind(decl->getKind());
  }
  static constexpr bool classofKind(Kind kind) {
    return kind == Kind::Method;
  }
};

class AnnotatedOperatorDecl final : public AnnotatedFunctionDecl {
public:
  AnnotatedOperatorDecl() : AnnotatedFunctionDecl(Kind::Operator) {}

  static bool classof(const AnnotatedDecl *decl) {
    return classofKind(decl->getKind());
  }
  static constexpr bool classofKind(Kind kind) {
    return kind == Kind::Operator;
  }
};

class AnnotatedConstructorDecl : public AnnotatedFunctionDecl {
public:
  bool implicit_conversion = false;

  AnnotatedConstructorDecl() : AnnotatedFunctionDecl(Kind::Constructor) {}

  bool processAnnotation(const clang::Decl *decl,
                         const annotations::Annotation &annotation) override;

  static bool classof(const AnnotatedDecl *decl) {
    return classofKind(decl->getKind());
  }
  static constexpr bool classofKind(Kind kind) {
    return kind == Kind::Constructor;
  }
};

class AnnotatedFieldOrVarDecl : public AnnotatedNamedDecl {
public:
  bool readonly = false;
  const clang::LambdaExpr *manual_bindings = nullptr;
  bool postamble = false;

  AnnotatedFieldOrVarDecl() : AnnotatedNamedDecl(Kind::Variable) {}

  bool processAnnotation(const clang::Decl *decl,
                         const annotations::Annotation &annotation) override;

  static bool classof(const AnnotatedDecl *decl) {
    return classofKind(decl->getKind());
  }
  static constexpr bool classofKind(Kind kind) {
    return kind == Kind::Variable;
  }
};

/// Holds and owns annotations s.t. they do not have to be processed multiple
/// times.  Currently only named declarations are checked for annotations.
class AnnotationStorage {
  llvm::DenseMap<const clang::NamedDecl *, std::unique_ptr<AnnotatedDecl>>
      annotations;

public:
  /// Add an entry for the specified `declaration`.
  /// If `declaration` is `nullptr` or unnamed, `nullptr` is returned.
  AnnotatedDecl *getOrInsert(const clang::Decl *declaration);

  /// Return the entry for the specified `declaration`.
  /// If `declaration` is `nullptr`, unnamed or there is no corresponding entry,
  /// `nullptr` is returned.
  const AnnotatedDecl *get(const clang::Decl *declaration) const;
};

} // namespace genpybind
