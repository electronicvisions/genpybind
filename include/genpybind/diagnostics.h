#pragma once

#include <clang/AST/Decl.h>
#include <clang/Basic/Diagnostic.h>
#include <clang/Basic/SourceLocation.h>

namespace genpybind {

class Diagnostics {
  clang::DiagnosticsEngine &engine;

public:
  enum class Kind {
    AlreadyExposedElsewhereError,
    AnnotationInvalidForDeclKindError,
    AnnotationInvalidSpellingError,
    AnnotationWrongArgumentTypeError,
    AnnotationWrongNumberOfArgumentsError,
    ConflictingAnnotationsError,
    ExposeHereCycleError,
    IgnoringQualifiersOnAliasWarning,
    PreviouslyExposedHereNote,
    UnreachableDeclContextWarning,
    UnsupportedAliasTargetError,
  };

  Diagnostics(clang::DiagnosticsEngine &engine) : engine(engine) {}

  static clang::DiagnosticBuilder report(const clang::Decl *decl, Kind kind);
  clang::DiagnosticBuilder report(clang::SourceLocation loc, Kind kind);
};

} // namespace genpybind
