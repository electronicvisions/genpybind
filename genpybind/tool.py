from __future__ import print_function
from __future__ import unicode_literals

import argparse
import os
import shutil
import subprocess
import sys
import tempfile

from clang.cindex import TranslationUnit

from .decls import gather_declarations
from .expose import expose_as

def main():
    # type: () -> None
    parser = argparse.ArgumentParser()

    parser.add_argument("--genpybind-module", dest="module", required=True)
    parser.add_argument("--genpybind-parse", dest="executable", default="genpybind-parse")
    parser.add_argument("--genpybind-docstring", dest="docstring")
    parser.add_argument("--genpybind-include", nargs="+", dest="includes")
    parser.add_argument("--genpybind-isystem", nargs="+", dest="isystem")
    parser.add_argument("--genpybind-tag", nargs="+", dest="tags")
    parser.add_argument("--genpybind-from-ast", dest="from_ast")
    parser.add_argument("--genpybind-output-files", dest="output_files", required=True,
        help="Comma-separated list of output file names.")
    parser.add_argument('rest', nargs=argparse.REMAINDER)

    # args, rest_args = parser.parse_known_args()
    args = parser.parse_args()

    if args.from_ast:
        if args.rest:
            parser.error(
                "unexpected arguments with --genpybind-from-ast: {}".format(args.rest))
        translation_unit = TranslationUnit.from_ast_file(args.from_ast)
    else:
        name = tempfile.mkdtemp(prefix="genpybind")
        ast_file = os.path.join(name, "genpybind.ast")
        try:
            rest = args.rest[:]
            if rest[0] == "--":
                del rest[0]
            status = subprocess.call(
                [args.executable, "-output-file", ast_file] + rest, stdout=sys.stderr)
            if status != 0:
                parser.error("genpybind-parse returned status {} when called with\n{}".format(
                    status, rest))
            translation_unit = TranslationUnit.from_ast_file(ast_file)
        finally:
            shutil.rmtree(name)

    if translation_unit.diagnostics:
        for diag in translation_unit.diagnostics:
            print("//", diag.format())
            for diag_ in diag.children:
                print("//", "  ", diag_.format())

    toplevel_declarations = gather_declarations(translation_unit.cursor)

    expose_as(
        toplevel_declarations,
        module=args.module,
        doc=args.docstring,
        isystem=args.isystem,
        includes=args.includes,
        tags=args.tags,
        output_files=args.output_files.split(','),
    )
