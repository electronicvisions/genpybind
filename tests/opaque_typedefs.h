#pragma once

#include "genpybind.h"

// Note: This feature will likely be renamed or redesigned, see
// https://github.com/kljohann/genpybind/issues/24.

// TODO: Add tests where the destination of the typename is in a different
// namespace, scope and/or submodule.
// TODO: Add an example where the destination is a templated type

// `opaque(false)` forces the target to be exposed at its own place of
// definition, using its own name.  Any extra keywords specified on the typedef
// will be passed on to the underlying type.  If it has already been made
// visible by an explicit annotation, the extra keywords have no effect and the
// typedef will just be an alias (NOTE: this is likely to change in the future).

struct TargetForOpaqueFalse {};
typedef TargetForOpaqueFalse typedef_opaque_false GENPYBIND(opaque(false));

// Extra arguments to `GENPYBIND` are passed on to underlying type.
struct TargetForOpaqueFalseExtraKeywords {};
typedef TargetForOpaqueFalseExtraKeywords
    typedef_opaque_false_extra_keywords GENPYBIND(opaque(false), dynamic_attr);

// Target already exposed: Extra arguments to `GENPYBIND` have no effect.
struct GENPYBIND(visible) TargetForOpaqueFalseAlreadyExposed {};
typedef TargetForOpaqueFalseAlreadyExposed
    typedef_opaque_false_already_exposed GENPYBIND(opaque(false), dynamic_attr);

// `opaque` or `opaque(true)` forces the target to be exposed at the location
// of the typedef (using the spelling / `expose_as` of the typedef).  If it has
// already been made visible by an explicit annotation or exposed elsewhere an
// error will be emitted during code generation.

// Note: This is the same as `opaque(true)`.
struct TargetForOpaque { int x = 1; };
typedef TargetForOpaque typedef_opaque GENPYBIND(opaque);

struct TargetForOpaqueTrue { int x = 2; };
typedef TargetForOpaqueTrue typedef_opaque_true GENPYBIND(opaque(true));

// Extra arguments to `GENPYBIND` are passed on to underlying type.
struct TargetForOpaqueTrueExtraKeywords {};
typedef TargetForOpaqueTrueExtraKeywords
    typedef_opaque_true_extra_keywords GENPYBIND(opaque(true), dynamic_attr);

// Error: It's not allowed to use `opaque(true)` if the target has already been
// exposed elsewhere.
// struct GENPYBIND(visible) TargetForOpaqueTrueAlreadyExposed {};
// typedef TargetForOpaqueTrueAlreadyExposed
//     typedef_opaque_true_already_exposed GENPYBIND(opaque(true));
