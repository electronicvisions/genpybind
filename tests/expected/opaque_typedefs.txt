NAME
    pyopaque_typedefs

CLASSES
    pybind11_builtins.pybind11_object(builtins.object)
        TargetForOpaqueFalse
        TargetForOpaqueFalseAlreadyExposed
        TargetForOpaqueFalseExtraKeywords
        typedef_opaque
        typedef_opaque_true
        typedef_opaque_true_extra_keywords

    class TargetForOpaqueFalse(pybind11_builtins.pybind11_object)
     |  Method resolution order:
     |      TargetForOpaqueFalse
     |      pybind11_builtins.pybind11_object
     |      builtins.object
     |
     |  Methods defined here:
     |
     |  __init__(...)
     |      __init__(*args, **kwargs)
     |      Overloaded function.
     |
     |      1. __init__(self: pyopaque_typedefs.TargetForOpaqueFalse) -> None
     |
     |      2. __init__(self: pyopaque_typedefs.TargetForOpaqueFalse, arg0: pyopaque_typedefs.TargetForOpaqueFalse) -> None
     |
     |  ----------------------------------------------------------------------
     |  Static methods inherited from pybind11_builtins.pybind11_object:
     |
     |  __new__(*args, **kwargs) from pybind11_builtins.pybind11_type
     |      Create and return a new object.  See help(type) for accurate signature.

    class TargetForOpaqueFalseAlreadyExposed(pybind11_builtins.pybind11_object)
     |  Method resolution order:
     |      TargetForOpaqueFalseAlreadyExposed
     |      pybind11_builtins.pybind11_object
     |      builtins.object
     |
     |  Methods defined here:
     |
     |  __init__(...)
     |      __init__(*args, **kwargs)
     |      Overloaded function.
     |
     |      1. __init__(self: pyopaque_typedefs.TargetForOpaqueFalseAlreadyExposed) -> None
     |
     |      2. __init__(self: pyopaque_typedefs.TargetForOpaqueFalseAlreadyExposed, arg0: pyopaque_typedefs.TargetForOpaqueFalseAlreadyExposed) -> None
     |
     |  ----------------------------------------------------------------------
     |  Static methods inherited from pybind11_builtins.pybind11_object:
     |
     |  __new__(*args, **kwargs) from pybind11_builtins.pybind11_type
     |      Create and return a new object.  See help(type) for accurate signature.

    class TargetForOpaqueFalseExtraKeywords(pybind11_builtins.pybind11_object)
     |  Method resolution order:
     |      TargetForOpaqueFalseExtraKeywords
     |      pybind11_builtins.pybind11_object
     |      builtins.object
     |
     |  Methods defined here:
     |
     |  __init__(...)
     |      __init__(*args, **kwargs)
     |      Overloaded function.
     |
     |      1. __init__(self: pyopaque_typedefs.TargetForOpaqueFalseExtraKeywords) -> None
     |
     |      2. __init__(self: pyopaque_typedefs.TargetForOpaqueFalseExtraKeywords, arg0: pyopaque_typedefs.TargetForOpaqueFalseExtraKeywords) -> None
     |
     |  ----------------------------------------------------------------------
     |  Data descriptors defined here:
     |
     |  __dict__
     |
     |  ----------------------------------------------------------------------
     |  Static methods inherited from pybind11_builtins.pybind11_object:
     |
     |  __new__(*args, **kwargs) from pybind11_builtins.pybind11_type
     |      Create and return a new object.  See help(type) for accurate signature.

    class typedef_opaque(pybind11_builtins.pybind11_object)
     |  Method resolution order:
     |      typedef_opaque
     |      pybind11_builtins.pybind11_object
     |      builtins.object
     |
     |  Methods defined here:
     |
     |  __init__(...)
     |      __init__(*args, **kwargs)
     |      Overloaded function.
     |
     |      1. __init__(self: pyopaque_typedefs.typedef_opaque) -> None
     |
     |      2. __init__(self: pyopaque_typedefs.typedef_opaque, arg0: pyopaque_typedefs.typedef_opaque) -> None
     |
     |  ----------------------------------------------------------------------
     |  Data descriptors defined here:
     |
     |  x
     |
     |  ----------------------------------------------------------------------
     |  Static methods inherited from pybind11_builtins.pybind11_object:
     |
     |  __new__(*args, **kwargs) from pybind11_builtins.pybind11_type
     |      Create and return a new object.  See help(type) for accurate signature.

    typedef_opaque_false = class TargetForOpaqueFalse(pybind11_builtins.pybind11_object)
     |  Method resolution order:
     |      TargetForOpaqueFalse
     |      pybind11_builtins.pybind11_object
     |      builtins.object
     |
     |  Methods defined here:
     |
     |  __init__(...)
     |      __init__(*args, **kwargs)
     |      Overloaded function.
     |
     |      1. __init__(self: pyopaque_typedefs.TargetForOpaqueFalse) -> None
     |
     |      2. __init__(self: pyopaque_typedefs.TargetForOpaqueFalse, arg0: pyopaque_typedefs.TargetForOpaqueFalse) -> None
     |
     |  ----------------------------------------------------------------------
     |  Static methods inherited from pybind11_builtins.pybind11_object:
     |
     |  __new__(*args, **kwargs) from pybind11_builtins.pybind11_type
     |      Create and return a new object.  See help(type) for accurate signature.

    typedef_opaque_false_already_exposed = class TargetForOpaqueFalseAlreadyExposed(pybind11_builtins.pybind11_object)
     |  Method resolution order:
     |      TargetForOpaqueFalseAlreadyExposed
     |      pybind11_builtins.pybind11_object
     |      builtins.object
     |
     |  Methods defined here:
     |
     |  __init__(...)
     |      __init__(*args, **kwargs)
     |      Overloaded function.
     |
     |      1. __init__(self: pyopaque_typedefs.TargetForOpaqueFalseAlreadyExposed) -> None
     |
     |      2. __init__(self: pyopaque_typedefs.TargetForOpaqueFalseAlreadyExposed, arg0: pyopaque_typedefs.TargetForOpaqueFalseAlreadyExposed) -> None
     |
     |  ----------------------------------------------------------------------
     |  Static methods inherited from pybind11_builtins.pybind11_object:
     |
     |  __new__(*args, **kwargs) from pybind11_builtins.pybind11_type
     |      Create and return a new object.  See help(type) for accurate signature.

    typedef_opaque_false_extra_keywords = class TargetForOpaqueFalseExtraKeywords(pybind11_builtins.pybind11_object)
     |  Method resolution order:
     |      TargetForOpaqueFalseExtraKeywords
     |      pybind11_builtins.pybind11_object
     |      builtins.object
     |
     |  Methods defined here:
     |
     |  __init__(...)
     |      __init__(*args, **kwargs)
     |      Overloaded function.
     |
     |      1. __init__(self: pyopaque_typedefs.TargetForOpaqueFalseExtraKeywords) -> None
     |
     |      2. __init__(self: pyopaque_typedefs.TargetForOpaqueFalseExtraKeywords, arg0: pyopaque_typedefs.TargetForOpaqueFalseExtraKeywords) -> None
     |
     |  ----------------------------------------------------------------------
     |  Data descriptors defined here:
     |
     |  __dict__
     |
     |  ----------------------------------------------------------------------
     |  Static methods inherited from pybind11_builtins.pybind11_object:
     |
     |  __new__(*args, **kwargs) from pybind11_builtins.pybind11_type
     |      Create and return a new object.  See help(type) for accurate signature.

    class typedef_opaque_true(pybind11_builtins.pybind11_object)
     |  Method resolution order:
     |      typedef_opaque_true
     |      pybind11_builtins.pybind11_object
     |      builtins.object
     |
     |  Methods defined here:
     |
     |  __init__(...)
     |      __init__(*args, **kwargs)
     |      Overloaded function.
     |
     |      1. __init__(self: pyopaque_typedefs.typedef_opaque_true) -> None
     |
     |      2. __init__(self: pyopaque_typedefs.typedef_opaque_true, arg0: pyopaque_typedefs.typedef_opaque_true) -> None
     |
     |  ----------------------------------------------------------------------
     |  Data descriptors defined here:
     |
     |  x
     |
     |  ----------------------------------------------------------------------
     |  Static methods inherited from pybind11_builtins.pybind11_object:
     |
     |  __new__(*args, **kwargs) from pybind11_builtins.pybind11_type
     |      Create and return a new object.  See help(type) for accurate signature.

    class typedef_opaque_true_extra_keywords(pybind11_builtins.pybind11_object)
     |  Method resolution order:
     |      typedef_opaque_true_extra_keywords
     |      pybind11_builtins.pybind11_object
     |      builtins.object
     |
     |  Methods defined here:
     |
     |  __init__(...)
     |      __init__(*args, **kwargs)
     |      Overloaded function.
     |
     |      1. __init__(self: pyopaque_typedefs.typedef_opaque_true_extra_keywords) -> None
     |
     |      2. __init__(self: pyopaque_typedefs.typedef_opaque_true_extra_keywords, arg0: pyopaque_typedefs.typedef_opaque_true_extra_keywords) -> None
     |
     |  ----------------------------------------------------------------------
     |  Data descriptors defined here:
     |
     |  __dict__
     |
     |  ----------------------------------------------------------------------
     |  Static methods inherited from pybind11_builtins.pybind11_object:
     |
     |  __new__(*args, **kwargs) from pybind11_builtins.pybind11_type
     |      Create and return a new object.  See help(type) for accurate signature.

